lib/hsa/mcwamp_hsa.cpp

```diff
@@ -1442,19 +1445,27 @@ class HSAQueue final : public KalmarQueue
 
 
 
-        if (!drainingQueue_ && (asyncOps.size() >= MAX_INFLIGHT_COMMANDS_PER_QUEUE-1)) {
+        if (!drainingQueueFailed_ && !drainingQueue_ && (asyncOps.size() >= HCC_MAX_INFLIGHT_COMMANDS_PER_QUEUE-1)) {
             DBOUT(DB_WAIT, "*** Hit max inflight ops asyncOps.size=" << asyncOps.size() << ". " << op << " force sync\n");
-            DBOUT(DB_RESOURCE, "*** Hit max inflight ops asyncOps.size=" << asyncOps.size() << ". " << op << " force sync\n");
+            DBOUT(DB_RESOURCE, "asyncOps=" << &asyncOps << " *** Hit max inflight ops asyncOps.size=" << asyncOps.size() << ". " << op << " force sync\n");
 
             drainingQueue_ = true;
 
             wait();
+
+            drainingQueue_ = false;
+
+            if (drainingQueueFailed_) {
+                /* draining queue failed, so truly force a wait */
+                DBOUTL(DB_RESOURCE, "draining queue failed, truly forcing sync");
+                wait();
+                drainingQueueFailed_ = false;
+            }
         }
-        op->asyncOpsIndex(asyncOps.size());
+        op->asyncOpsIndex(asyncOps.size()+asyncOps_offset);
```

Comment:
any statement be associated with `asyncOps_offset` deserves to have comment on the effect brought by this important field.